File: apps/api/pencheff_api/schemas/llm_providers.py
from __future__ import annotations

from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field, model_validator

from ..services.llm_providers.catalog import PROVIDER_KINDS

LlmProviderKind = Literal[
    "openai", "anthropic", "google", "azure_openai", "openai_compatible"
]


def _validate_kind_fields(provider: str, base_url: str | None,
                          azure_deployment: str | None,
                          azure_api_version: str | None) -> None:
    if provider not in PROVIDER_KINDS:
        raise ValueError(f"unknown provider {provider!r}")
    if provider in ("openai_compatible", "azure_openai") and not base_url:
        raise ValueError(f"{provider} requires base_url")
    if provider == "azure_openai" and not (azure_deployment and azure_api_version):
        raise ValueError("azure_openai requires azure_deployment + azure_api_version")


class LlmProviderCreate(BaseModel):
    label: str = Field(min_length=1, max_length=120)
    provider: LlmProviderKind
    model: str = Field(min_length=1, max_length=200)
    base_url: str | None = Field(default=None, max_length=1024)
    # Required for hosted providers; openai_compatible may use a keyless
    # self-host, so an empty key is allowed only there.
    api_key: str = ""
    azure_deployment: str | None = Field(default=None, max_length=200)
    azure_api_version: str | None = Field(default=None, max_length=40)
    extra: dict | None = None

    @model_validator(mode="after")
    def _check(self) -> "LlmProviderCreate":
        _validate_kind_fields(self.provider, self.base_url,
                              self.azure_deployment, self.azure_api_version)
        if self.provider != "openai_compatible" and not self.api_key:
            raise ValueError(f"{self.provider} requires a non-empty api_key")
        return self


class LlmProviderUpdate(BaseModel):
    label: str | None = Field(default=None, min_length=1, max_length=120)
    provider: LlmProviderKind | None = None
    model: str | None = Field(default=None, min_length=1, max_length=200)
    base_url: str | None = Field(default=None, max_length=1024)
    # None = leave unchanged; "" = clear the stored key; non-empty = replace.
    api_key: str | None = None
    azure_deployment: str | None = Field(default=None, max_length=200)
    azure_api_version: str | None = Field(default=None, max_length=40)
    extra: dict | None = None


class LlmProviderOut(BaseModel):
    id: str
    label: str
    provider: str
    model: str
    base_url: str | None = None
    azure_deployment: str | None = None
    azure_api_version: str | None = None
    extra: dict | None = None
    key_set: bool = False
    key_hint: str | None = None  # e.g. "…AB12"
    is_active: bool = False
    created_at: datetime

    model_config = {"from_attributes": True}

File: apps/web/lib/llm-providers.ts
import { api } from "./api";

export type LlmProviderKind =
  | "openai"
  | "anthropic"
  | "google"
  | "azure_openai"
  | "openai_compatible";

export type LlmProvider = {
  id: string;
  label: string;
  provider: LlmProviderKind;
  model: string;
  base_url?: string | null;
  azure_deployment?: string | null;
  azure_api_version?: string | null;
  extra?: Record<string, unknown> | null;
  key_set: boolean;
  key_hint?: string | null;
  is_active: boolean;
  created_at: string;
};

export type LlmProviderInput = {
  label: string;
  provider: LlmProviderKind;
  model: string;
  base_url?: string;
  api_key?: string; // omit on edit to keep unchanged; "" to clear
  azure_deployment?: string;
  azure_api_version?: string;
};

export const listProviders = () => api<LlmProvider[]>("/llm-providers");
export const getCatalog = () =>
  api<{
    kinds: LlmProviderKind[];
    models: Record<string, { id: string; label: string }[]>;
  }>("/llm-providers/catalog");
export const createProvider = (body: LlmProviderInput) =>
  api<LlmProvider>("/llm-providers", { method: "POST", json: body });
export const updateProvider = (id: string, body: Partial<LlmProviderInput>) =>
  api<LlmProvider>(`/llm-providers/${id}`, { method: "PATCH", json: body });
export const deleteProvider = (id: string) =>
  api<void>(`/llm-providers/${id}`, { method: "DELETE" });
export const activateProvider = (id: string) =>
  api<LlmProvider>(`/llm-providers/${id}/activate`, { method: "POST" });
export const deactivateProvider = () =>
  api<void>("/llm-providers/deactivate", { method: "POST" });
export const testProvider = (id: string) =>
  api<{
    ok: boolean;
    latency_ms: number;
    error: string | null;
    model: string;
    sample?: string;
  }>(`/llm-providers/${id}/test`, { method: "POST" });
